### DISCUSSION TITLE
Web export

### DISCUSSION THREAD

**Discussion by SimonBiggs** [lexiconhq/lexicon#38] [Ideas]
Hi,

I'd be quite keen to undergo the required work to make it so that lexicon is also able to export to a webpage as well.

Might I be able to have access to the current branch where you're doing work internally so that I can investigate the plausibility of achieving this?

Cheers,
Simon

**Comment by Kinostrome** [comment]
Hi @SimonBiggs. Thanks for your interest.

One of the main purposes of Lexicon is specifically that it _is not_ a web page, but provides a seamless native experience, and focuses solely on that. Please refer to the [Background & Motivation](https://docs.lexicon.is/rationale) for a better understanding of the goals of this project.

This is because Discourse already provides a fully functional mobile website. Their mobile site is what is used in their official [DiscourseHub](https://apps.apple.com/us/app/discourse-hub/id1173672076) mobile app.

If you just want Discourse's mobile site to _look_ more like Lexicon on mobile, we would encourage you to explore creating a [custom Discourse theme](https://meta.discourse.org/t/beginners-guide-to-developing-discourse-themes/93648).

**Reply by SimonBiggs** [reply]
Hi @Kinostrome, we have been undergoing theming with Discourse for the last 1 1/2 months. A key issue is the feature miss-match between Lexicon and the Discourse webpage. For example, chat not being implemented within Lexicon. As such, we want to cause an effective limitation on what is possible within the web version, so that people who are using the app only don't end up missing out on things that have been sent to them through the web version.

We also really appreciate the design work that has already been undergone within Lexicon itself.

Also, for the particular use case we are targeting, having the browser version "feel like an app" will not be an issue. At the end of the day, the majority of users will be via the app.

---

If I did undergo the work to include all of the expo web component requirements to have lexicon build to the web, is that something that you would not be interested in merging?

---

As a note, community managers would still regularly utilise the Discourse vanilla webpage in order to manage settings and other admin tasks. But, the purpose is that every non-admin user will be funnelled through to the Lexicon interface, and use it only, leaning in on its simplicity.

**Reply by Kinostrome** [reply]
Thanks for the explanation @SimonBiggs.

We do generally want to keep the scope of Lexicon relatively focused. We're initially apprehensive just because any change like that can add quite a bit of complexity—even if it seems simple at first—which later makes the project harder to maintain since it's supporting multiple use cases.

We'd certainly be open to reviewing a possible implementation, but it's hard to say how that would go. For that reason it may be worth submitting a plan or a proposal for how such changes would be implemented, motivation, goals, impact, etc. This could include planned areas / files for refactor (if needed), how the build process would work, etc.

That would be less of an investment on your end and would give a better idea of how it may affect the project scope.

An alternative might be to fork and try it out first to see how it goes, so you aren't so blocked on us and can get an idea if it's feasible, what the downsides of that approach might be, etc.

**Reply by SimonBiggs** [reply]
> That would be less of an investment on your end and would give a better idea of how it may affect the project scope.

Yup makes sense. I think on our end, for now, we'll play wait and see until the dust settles in the Lexicon repo. Once the open source repo is your primary source of truth I'll revisit the best course of action.

**Reply by SimonBiggs** [reply]
> which later makes the project harder to maintain since it's supporting multiple use cases.

Out of interest @Kinostrome, if as a part of every PR I provided I also made sure that there was an appropriate GitHub CI activity that was utilising the web-page converted features in an end-to-end fashion utilising https://playwright.dev/, might that be something that helps with maintainability?

I would also make sure to follow the "small PR" moto (https://google.github.io/eng-practices/review/developer/small-cls.html), combined with thorough testing in each PR, hopefully this would make code review on your end easier to manage... And, if at any point the direction isn't the path you wanted then the PR would be small in the end, and the work lost would be small...

---

Practically, to achieve this, I would make the porting iterative, setting up a separate web entry point, and only slowly, piece-by-piece move over the conversion.

**Reply by Kinostrome** [reply]
That approach sounds wise and would be generally helpful for any contributions.

For this particular proposed set of changes, we would still defer to the previous point about proving it out on a fork for a bit first.

Keep in mind that all usages of Lexicon are forks anyway—because each Lexicon-powered mobile app must be customized for a particular Discourse site. So that approach is fairly standard regardless of the changes being made.

Since this project is mobile-app-centric, we are currently building out an implementation of testing with [Detox](https://wix.github.io/Detox/) testing. That is the main priority because we primarily want to catch bugs in the mobile app itself, rather than an exported webpage version.

**Reply by SimonBiggs** [reply]
Yup, makes sense.

> For this particular proposed set of changes, we would still defer to the previous point about proving it out on a fork for a bit first.

Okay, looks like that's the pathway forward. I'll come back to you once we have something a bit more proved out initially.

Thanks @Kinostrome,
Cheers,
Simon

**Reply by SimonBiggs** [reply]
Would you recommend I branch off of `alpha-v2` or `master` for our dev work? (eg, if we have fly-by bug fixes that are not related to the web export, which of those branches would most benefit you for them to come to?)

**Reply by Kinostrome** [reply]
That depends specifically on when and how you're intending to use it.

`master` won't have many feature changes (and won't include things like push notifications, integration with the upcoming Lexicon Discourse plugin, etc.), but will be true to the experience that's currently presented in the documentation for v1. And, as bugfixes and issues get resolved, it can return to a more stable state with _only_ fixes coming in (rather than new features).

`alpha-v2` by nature will be less stable, but will include new features, and eventually will become the active development branch.

If there is a bug fix that would benefit both branches, we would probably end up applying it to both. There are of course some exceptions to that, such as a bugfix that would end up pulling v2 changes into v1 or complicating things in that way. But that gets handled on a case-by-case basis.

**Reply by SimonBiggs** [reply]
> That depends specifically on when and how you're intending to use it.

We're looking for an initial deployment to be pegged in September.

**Reply by SimonBiggs** [reply]
@Kinostrome, here's an update. I have the application exporting to a statically generated webpage using GitHub pages over here:
https://lexicon.guildofentrepreneurs.com/

The same code base still successfully builds to an app and passes all the mobile tests.

I had to disable icons to begin with when the platform is web. And there is still a few bugs on the web side. But, due to the use of platform flags, there is a potential to be able to create a minimal PR that merges the initial web export prototype if this is something you might be interested in.

Let me know :) (I'll go fix up the other PRs now)

**Reply by Kinostrome** [reply]
Nice, that looks cool. There does appear to already be a bug related to the animation with the search bar header by the way.

For now, we'd prefer the team focus on the upcoming beta release, so it is best maintained on a fork. Since, as you're saying, it's a small set of changes, keeping the fork in sync with the changes here doesn't sound like it would be too much of an issue.
